### ISSUE TITLE
Doesn't work with e10s

### ISSUE THREAD

**Issue by medopaw** [yxl/Fire-IE#128]
Fails to load any page with IE engine in latest nightly with [e10s](https://wiki.mozilla.org/Electrolysis) on.

Also checkout http://arewee10syet.com/

**Comment by patwonder** [comment]
Manual switching is working as of b3eddd1bfb77039de2570b648f39be921273db80. Auto-switch does not work yet because content policy is processed in content processes, not the chrome process.

**Comment by patwonder** [comment]
Auto-switch should mostly work as of dbd7968cf2d52bb52d6e1c7b913ef6fa9fac5891. There is a minor issue though:
* Auto-switch-back in a new window results in the throbber rotating infinitely (without showing actual content). The content is actually loaded and hovering over (invisible) links changes the cursor. Might be a Nightly bug.

**Comment by patwonder** [comment]
e10s branch merged into dev.

**Comment by patwonder** [comment]
e10s support has been merged into master branch. Consider issue fixed unless a future Nightly version breaks anything.

**Comment by patwonder** [comment]
Reopen because a recent patch in Nightly breaks plugin loading. Nightly purposely disabled plugin loading in the chrome process when e10s is enabled (see [Bug 1155976](https://bugzilla.mozilla.org/show_bug.cgi?id=1155976), comment 8). However, this patch makes no sense to me, as they are fixing something that will eventually come out (the non-e10s window menu).

A corresponding bug has been filed (hoping) to address this issue, see [Bug 1158788](https://bugzilla.mozilla.org/show_bug.cgi?id=1158788). If the nonsense patch is going to stay, we should at least have an option to disable it, or have our plugin white-listed.

@yxl any thoughts?

**Comment by yxl** [comment]
@patwonder I read through all the bug comments (either from here or mozilla bugzilla) and really appreciate what you have done! 
OOPP has been a pain for us for a long time. I think it's time for us to solve it directly without workarounds any more.
I'd like to create our own IPC tunnel between the plugin and the chrome process to replace the heavy IPC mechanism of the Gecko. I'm not sure which IPC method to use now. Socket may be an option.

**Comment by patwonder** [comment]
@yxl Yeah, creating our own IPC mechanism would be a good idea. It could be very lightweight as we only support limited IPC messages. Either named pipe or socket is OK.

I'm also thinking of getting rid of the NPAPI plugin, and use js-ctypes to load the IE control directly. But that requires us manually managing window positioning / visibility / resizing, as well as IE control's lifetime, which a NPAPI plugin pretty much does automatically.

**Comment by yxl** [comment]
>I'm also thinking of getting rid of the NPAPI plugin, and use js-ctypes to load the IE control directly. But that requires us manually managing window positioning / visibility / resizing, as well as IE control's lifetime, which a NPAPI plugin pretty much does automatically.

Another weak point of js-ctypes is that it is synchronous and doesn't support async callbacks. So we still need to establish own message channel with js-ctypes way.

What's the benefit of using js-ctypes?

**Comment by patwonder** [comment]
> What's the benefit of using js-ctypes?

Mozilla has been phasing out plugin usage for a long time: first click-to-play, then the plugin doorhangers, and recently the Nightly patch. It's a pain for us to keep up with Mozilla's changes, too, in addition to OOPP. Js-ctypes, on the other side, seems to be much more stable.
